### PR TITLE
Apply update track before next start

### PR DESCRIPTION
**Pull request by olmisa**
* [ ] Has `CHANGELOG.md` been updated?
* [ ] Are tests passing locally?
* [x] Are the files formatted correctly?
* [ ] Did you add unit tests if this modifies the Windows code?
* [x] Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

## Description

Possibility to update track before next app start.

## Related PRs or issues

https://github.com/microsoft/appcenter-sdk-unity/pull/395
[AB#75908](https://msmobilecenter.visualstudio.com/454a20a9-dfe6-4c1f-8ae8-417f76adb472/_workitems/edit/75908)

### CODE REVIEW

File: AppCenterLoaderApp/appcenter-loader/src/main/java/com/microsoft/appcenter/loader/AppCenterLoader.java [hunk 1]
```diff
@@ -33,6 +33,8 @@
     private static final String USE_CUSTOM_LOG_URL_KEY = "appcenter_use_custom_log_url";
     private static final String MAX_STORAGE_SIZE = "appcenter_max_storage_size";
     private static final String INITIAL_LOG_LEVEL_KEY = "appcenter_initial_log_level";
+    private static final String INITIAL_UPDATE_TRACK_KEY = "appcenter_initial_update_track";
+    private static final String INITIAL_WHEN_UPDATE_TRACK_KEY = "appcenter_initial_when_update_track";
```
Comment (guperrot) [inline AppCenterLoaderApp/appcenter-loader/src/main/java/com/microsoft/appcenter/loader/AppCenterLoader.java]: While this was fine for an internal test app UI, this is too confusing at SDK settings level. We agreed to combine those into a 3 option drop down, where the default value wouldn't call setUpdateTrack before start. You can find the poll for the third value in our team channel which should have votes by the time you see this tomorrow.

**Review by guperrot — CHANGES_REQUESTED** [review]

File: Assets/AppCenter/Editor/AppCenterSettingsMakerAndroid.cs [hunk 2]
```diff
@@ -49,6 +51,16 @@ public void SetStartupType(int startupType)
...
     }
 
+    public void SetUpdateTrack(int updateTrack)
+    {
+        _resourceValues[UpdateTrackKey] = updateTrack.ToString();
+    }
+
+    public void SetWhenUpdateTrack(int whenUpdateTrack)
+    {
+        _resourceValues[WhenUpdateTrackKey] =  whenUpdateTrack.ToString();
```
Comment (owlstronaut) [inline Assets/AppCenter/Editor/AppCenterSettingsMakerAndroid.cs]:
```suggestion
        _resourceValues[WhenUpdateTrackKey] = whenUpdateTrack.ToString();
```

File: AppCenterLoaderApp/appcenter-loader/src/main/java/com/microsoft/appcenter/loader/AppCenterLoader.java (hunk 1 above)
Comment (guperrot) [reply AppCenterLoaderApp/appcenter-loader/src/main/java/com/microsoft/appcenter/loader/AppCenterLoader.java]: So we will only allow before start. So name it only UPDATE_TRACK_KEY (and remove initial from value), and remove the timing.

**Review by guperrot — CHANGES_REQUESTED** [review]

**Comment by guperrot** [comment]
Also if the C# bindings become useless if they can be called only before start, we might want to remove them.

File: AppCenterLoaderApp/appcenter-loader/src/main/java/com/microsoft/appcenter/loader/AppCenterLoader.java (hunk 1 above)
Comment (olmisa) [reply AppCenterLoaderApp/appcenter-loader/src/main/java/com/microsoft/appcenter/loader/AppCenterLoader.java]: Fixed

**Review by annakocheshkova — APPROVED** [review]

**Comment by guperrot** [comment]
> Did you test your change with either the sample apps that are included in the repository or with a blank app that uses your change?

File: Assets/AppCenter/Plugins/AppCenterSDK/Distribute/iOS/DistributeInternal.cs [hunk 3]
```diff
@@ -19,7 +19,7 @@ class DistributeInternal
 
         static ReleaseAvailableDelegate del;
         static IntPtr ptr;
-
+
```
Comment (guperrot) [inline Assets/AppCenter/Plugins/AppCenterSDK/Distribute/iOS/DistributeInternal.cs]:
```suggestion

```

**Review by guperrot — CHANGES_REQUESTED** [review]
This does not work correctly on iOS.

Repro: use this branch and thus without the UI test changes (we need to verify with just the AppCenterBehavior change from editor).

Set up Unity to target a real device.
Change track to private in Editor via the behavior.
Build and run on a device.
Stop debugger.
Restart app from device without debugger.

Expected: browser opens.
Actual: checking for public update.

Storage seems to be nil:

Screen Shot 2020-02-04 at 4 36 08 PM

This works on Android though.

**Comment by guperrot** [comment]
By the way those methods are not called, do we need cleanup?

Screen Shot 2020-02-04 at 4 38 29 PM

File: Assets/AppCenter/Editor/AppCenterSettingsMakerIos.cs [hunk 4]
```diff
@@ -24,7 +24,8 @@ public class AppCenterSettingsMakerIos : IAppCenterSettingsMaker
     private const string InstallUrlToken = "APPCENTER_UNITY_USE_CUSTOM_INSTALL_URL";
     private const string UseCustomMaxStorageSize = "APPCENTER_USE_CUSTOM_MAX_STORAGE_SIZE";
     private const string MaxStorageSize = "APPCENTER_MAX_STORAGE_SIZE";
-
+    private const string UpdateTrackSearchText = "1/*UPDATE_TRACK*/";
```
Comment (guperrot) [inline Assets/AppCenter/Editor/AppCenterSettingsMakerIos.cs]: There is a space in the objective c file, I am confused.

File: Assets/AppCenter/Plugins/iOS/Core/AppCenterStarter.original [hunk 5]
```diff
@@ -94,6 +96,10 @@ static const int kMSStartupType = 0 /*STARTUP_TYPE*/;
 
 #ifdef APPCENTER_UNITY_USE_DISTRIBUTE
 
+NSNumber *updateTrack = [[NSUserDefaults standardUserDefaults] objectForKey:kMSUpdateTrackKey];
```
Comment (guperrot) [inline Assets/AppCenter/Plugins/iOS/Core/AppCenterStarter.original]: Why is this read from storage but the editor is just replacing a constant. If the editor is editing code I don't understand why storage is needed here.

Comment (olmisa) [reply Assets/AppCenter/Plugins/iOS/Core/AppCenterStarter.original]: It was done for the test app. But if we don't need test app changes then it should be deleted.

File: Assets/AppCenter/Editor/AppCenterSettingsMakerIos.cs (hunk 4 above)
Comment (olmisa) [reply Assets/AppCenter/Editor/AppCenterSettingsMakerIos.cs]: Fixed

**Review by guperrot — APPROVED** [review]

File: Assets/AppCenter/Editor/AppCenterSettingsMakerIos.cs (hunk 4 above)
Comment (guperrot) [reply Assets/AppCenter/Editor/AppCenterSettingsMakerIos.cs]: Interesting, do other settings work? The ones that use the same pattern.

**Pull request merged**